src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+/// <summary>
+/// An <see cref="IBufferWriter{T}"/> implementation for writing to
+/// unmanaged blocks of memory.
+/// </summary>
+internal sealed unsafe class UnmanagedBufferWriter : IBufferWriter<byte>, IDisposable
+{
+    private readonly int _blockSize;
+    private readonly List<IntPtr> _allocations = new();
+    private byte* _currentBlock;
+    private int _currentBlockCount;
+
+    /// <summary>
+    /// Instantiate an <see cref="UnmanagedBufferWriter"/> instance.
+    /// </summary>
+    /// <param name="blockSize">The unmanaged memory block size in bytes.</param>
+    public UnmanagedBufferWriter(int blockSize)
+    {
+        _blockSize = blockSize;
+        _currentBlockCount = -1;
+    }
+
+    /// <inheritdoc />
+    public void Advance(int count)
+    {
+        Debug.Assert(count >= 0);
+
+        _currentBlockCount -= count;
+        if (_currentBlockCount > 0)
+        {
+            _currentBlock += count;
+        }
+        else
+        {
+            NewBlock();
```

Comment:
This does not look right. Advance always only commit what was written. It should not need to ever allocate a new chunk.